kolibri/content/api.py

```diff
@@ -38,17 +45,109 @@ def title_description_filter(self, queryset, value):
             reduce(lambda x, y: x & y, token_queries))
 
     def filter_recommendations_for(self, queryset, value):
+        """
+        Recommend items that are similar to this piece of content.
+        """
         recc_node = queryset.get(pk=value)
         descendants = recc_node.get_descendants(include_self=False).exclude(kind__in=['topic', ''])
         siblings = recc_node.get_siblings(include_self=False).exclude(kind__in=['topic', ''])
         data = descendants | siblings  # concatenates different querysets
         return data
 
-    def filter_recommendations(self, queryset, value):
-        # return 25 random content nodes
-        pks = queryset.values_list('pk', flat=True).exclude(kind__in=['topic', ''])
-        count = min(pks.count(), 25)
-        return queryset.filter(pk__in=sample(list(pks), count))
+    def filter_next_steps(self, queryset, value):
+        """
+        Recommend uncompleted content, content that has user completed content as a prerequisite.
+
+        :param queryset: all content nodes for this channel
+        :param value: id of currently logged in user, or none if user is anonymous
+        :return: uncompleted content nodes, or empty queryset if user is anonymous
+        """
+
+        # if user is anonymous, don't return any nodes
+        if value == '':
+            return queryset.none()
+
+        if self.data['channel']:
+            from kolibri.content.content_db_router import using_content_database
+            from kolibri.content.models import ContentNode
+            with using_content_database(self.data['channel']):
+                # custom SQL query to get uncompleted content based on mptt algorithm
+                return ContentNode.objects.raw('''SELECT incomplete_node.*
```

Comment:
- [ ] Need to change this raw SQL query to dynamically take the database table names.